### ISSUE TITLE
Discrepency between 'Success' and actual predictions

### ISSUE THREAD

**Issue by KarthikGanesan88** [bethgelab/foolbox#674]
I'm seeing a small discrepancy between the number of images that successfully fooled the network when checking `success` vs. getting predictions for those images by running it through the model again. 

Unfortunately, I am only seeing it in my own code (which is a bit too big to separate into a MWE to upload here) and not in the [examples](https://github.com/bethgelab/foolbox/tree/master/examples). I'm running a PreActResNet18 model, with the cifar10 dataset. 

```
classifier = PyTorchModel(model, bounds=(-1.98, 2.12))
_, x_adv, success = FGSM(classifier, X, y, epsilons=epsilon)
y_pred = classifier(x_adv).argmax(axis=-1)
print(f'Success.sum(): {success.sum()}')
print(f'Pred.sum():{(y_pred != y).sum().item()}')
```
I am only iterating here over 100 inputs that the network correctly classified (i.e., accuracy for 'X' here is 100%). Given that, I would expect both to give me the same number. But for certain epsilon values, they are different: 

| Epsilon | 0.003 | 0.001  | 0.1 |
| ---- | ---| ---| ---|
| Success.sum() | 1 | 3 | 17 |
| Pred.sum() | 0 | 2 | 18 | 

Is there any reason they should be different? For other epsilon values (e.g., 0.005, 0.01, 0.02, 0.05) they are the same. 

Since I can't (easily) provide a MWE, here are some things that I saw are different between my code and the examples, in case the issue is caused by one of these:

1. I am not doing pre-processing by passing in a parameter to `PyTorchModel`. Rather, I am using PyTorch's `transforms.Normalize` to do this. 
2. My data is not bounded to (0,1) but rather to an odd bound of (-1.98, 2.12). Could this be causing this mis-match?

**Comment by zimmerrol** [comment]
Short explanation for people you might stumble over this issue: If your data is not in the `[0, 1]` bound you _need_ to pass the new bounds as the `bounds` argument when creating your `PyTorchModel`. Otherwise, the attack internally assumes it can only change data within the standard `[0, 1]` range, which means the attack _overestimates_ the robustness of your classifier.
Since you set the `bounds` argument in your example, you're doing everything correctly here:)

For me this looks more like the result you would expect from a non-deterministic classifier. Have you made sure that the model is set in eval mode? If it's still in train mode, its batch norm statistics are not kept fixed but change during the attack which affects the output of the classifier.

**Comment by KarthikGanesan88** [comment]
Hi @zimmerrol , thank you. That makes sense. There is something odd about the batchnorm in my model. I am using a pre-trained model that is doing some tricks with the batchnorm (by loading different values each time the model is loaded). But it should be totally deterministic. I also set the torch and numpy random seeds at the start to force determinism. But I am going to train a new model with a more straightforward batchnorm. If the issues goes away there, I will come back and reply here. Or if I am able to find a simpler model which shows this behaviour, I will reply with that to make it easier to find out why. 

I did have one other basic question if you don't mind me asking that here. Does Foolbox use the model's own backward functions for carrying out the attacks? So I am using a model that I am training myself (with custom forward and backward functions for Conv2D and BN layers). When I do the attacks, does it use those functions or does PyTorchModel use the 'standard' backward functions for Conv2D and BN?

**Comment by zimmerrol** [comment]
Foolbox just uses the backprop algorithms from the framework in use. This means, that if you use e.g., PyTorch and use a function with a custom backward pass, this will also be used to calculate the gradients for foolbox. Hope that helps.
